tempesta_fw/apm.c

```diff
@@ -0,0 +1,1041 @@
+			tfw_stats_adjust(rng, 1);
+			atomic64_inc(&rng->total_cnt);
+			return;
+		}
+		atomic_inc(__rng(&pc0, rng->cnt[0], r_time));
+		tfw_stats_adjust(rng, 0);
+		atomic64_inc(&rng->total_cnt);
+		return;
+	}
+
+	if (!spin_trylock(&tfw_sa_guard)) {
+		TFW_INC_STAT_BH(apm.upd_misses);
+		return;
+	}
+	pc3.atomic = rng->ctl[3].atomic;
+	if (unlikely(r_time > pc3.end)) {
+		tfw_stats_extend(rng, r_time);
+		pc3.atomic = rng->ctl[3].atomic;
+	}
+	atomic_inc(__rng(&pc3, rng->cnt[3], r_time));
+	__tfw_stats_adjust(rng, 3);
+	atomic64_inc(&rng->total_cnt);
+	spin_unlock(&tfw_sa_guard);
+}
+
+/*
+ * APM ring buffer.
+ *
+ * It consists of the predefined number of entries that are "reused" as
+ * the buffer gets full. The ring buffer as a whole keeps the APM stats
+ * for the latest time interval (the time window), and each entry of
+ * the buffer keeps the APM stats for a piece of that time (an interval).
+ */
+/*
+ * A ring buffer entry structure.
+ * @pcntrng	- Struct for response time data by the percentiles algorithm.
+ * @jtmistamp	- The time the ring buffer entry has been started, in jiffies.
+ * @reset	- The entry can be reset by one thread at a time.
+ * @lock	- The lock to adjust the ranges of response times.
+ */
+typedef struct {
+	TfwPcntRanges	pcntrng;
+	unsigned long	jtmistamp;
+	atomic_t	reset;
+} TfwApmRBEnt;
+
+/*
+ * The ring buffer contol structure.
+ * This is a supporting structure. It keeps related data that is useful
+ * in making decisions on the need of recalculation of percentiles.
+ * @jtmwstamp	- The start of the time window the percentiles are for.
+ * @entry_cnt	- The number of hits in the current buffer ring entry.
+ * @total_cnt	- The number of hits within the current time window.
+ */
+
+typedef struct {
+	unsigned long	jtmwstamp;
+	unsigned long	entry_cnt;
+	unsigned long	total_cnt;
+} TfwApmRBCtl;
+
+/*
+ * The ring buffer structure.
+ * @rbent	- Array of ring buffer entries.
+ * @rbufsz	- The size of @rbent.
+ */
+typedef struct {
+	TfwApmRBEnt	*rbent;
+	int		rbufsz;
+} TfwApmRBuf;
+
+/*
+ * The percentiles data structure.
+ * Keeps the latest calculated percentiles.
+ * @percentile	- Array of percentiles.
+ * @prcntlsz	- The size of @percentile.
+ * @rwlock	- Protect updates.
+ */
+typedef struct {
+	PrcntlStats	*pstats;
+	unsigned int	pstsz;
+	rwlock_t	rwlock;
+} TfwApmPrcntl;
+
+/*
+ * The stats data structure.
+ * There's only one updater that runs on timer and calculates the latest
+ * percentiles. There may be multiple readers of the latest percentiles.
+ * Readers read the latest percentiles at @prcntl[@rdidx % 2]. The writer
+ * writes the new percentiles to @prcntl[(@rdidx + 1) % 2], and then
+ * increments @rdidx. The reading and writing is protected by a rwlock.
+ * @prcntl	- The percentiles for reading/writing (flip-flop manner).
+ * @rdidx	- The current index in @prcntl for readers.
+ */
+typedef struct {
+	TfwApmPrcntl	prcntl[2];
+	atomic_t	rdidx;
+} TfwApmStats;
```

Comment:
Oh, my failure - I forgot that the lock contention happens only for slow readers, which isn't likely. So please just describe the locking solution in the comment above.